streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -0,0 +1,428 @@
+                //if we've already seen the window with the closest start time to the record
+                boolean foundFirst = false;
+                //if we've already seen the window with the closest end time to the record
+                boolean foundFirstEndTime = false;
+                while (iterator.hasNext()) {
+                    next = iterator.next();
+                    windowStartTimes.add(next.key.window().start());
+
+                    //determine if current record's right window exists
+                    if (next.key.window().start() == timestamp + 1) {
+                        rightWindowExists = true;
+                        continue;
+                    } else if (next.key.window().end() > timestamp) {
+                        if (!foundFirst) {
+                            foundFirst = true;
+                            if (isLeftWindow(next)) {
+                                foundLeftFirst = true;
+                            }
+                        }
+                        rightWinAgg = next.value;
+                        putAndForward(next.key.window(), next.value, key, value, closeTime, timestamp);
+                        continue;
+                    } else if (next.key.window().end() == timestamp) {
+                        putAndForward(next.key.window(), next.value, key, value, closeTime, timestamp);
+                        leftWinExists = true;
+                        continue;
+                    } else {
+                        if (!foundFirstEndTime) {
+                            leftWinAgg = next.value;
+                            foundFirstEndTime = true;
+                        }
+                        if (prevRightWinExists) {
+                            break;
+                        }
+                        if (isLeftWindow(next)) {
+                            prevRightWinExists = true;
+                            final long rightWinStart = next.key.window().end() + 1;
+                            if (windowStartTimes.contains(rightWinStart)) {
+                                prevRightWinAlreadyCreated = true;
+                            } else {
+                                final TimeWindow window = new TimeWindow(rightWinStart, rightWinStart + windows.timeDifference);
+                                final ValueAndTimestamp<Agg> valueAndTime = ValueAndTimestamp.make(initializer.apply(), timestamp);
+                                putAndForward(window, valueAndTime, key, value, closeTime, timestamp);
+                            }
+                        }
+                    }
+                }
+            }
+
+            //create the left window of the current record if it's not created
+            if (!leftWinExists) {
+                final Agg aggValue;
+                final long newTimestamp;
+                //confirms that the left window contains more than the current record
+                if (prevRightWinExists) {
```

Comment:
Ok I may have lost the trail of logic here...are we just checking `prevRightWinExists` as an indicator of whether we actually found any records to the left of our record within range? Could/should we check `foundFirstEndTime` instead?

Reply:
Yeah that's what `prevRightWinExists` is doing right now, we could store the full `valueAndTimestamp` for `foundFirstEndTime` and check to see if the max timestamp is within the range of recordTime-timeDifference

Reply:
changed to
`                if (leftWinAgg.timestamp() < timestamp && leftWinAgg.timestamp() > timestamp - windows.timeDifferenceMs()) {
                    valueAndTime = ValueAndTimestamp.make(leftWinAgg.value(), timestamp);
                } else {
                    //left window just contains the current record
                    valueAndTime = ValueAndTimestamp.make(initializer.apply(), timestamp);
                }`